File: 101SymmetricTree.cpp
//1. Recursion. O(N)
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(!root) return true;
        return isSymmetric(root->left,root->right);
    }
    
    bool isSymmetric(TreeNode* p,TreeNode* q) {
        if(!p && q || p && !q) return false;
        if(!p && !q) return true;
        if(p && q) return (p->val==q->val) && isSymmetric(p->left,q->right) && isSymmetric(q->left,p->right);
    }
};

//2. Iteration. O(n)
class Solution{
public:
    bool isSymmetric(TreeNode *root) {
        if(!root) return true;
        stack<TreeNode*> s;
        s.push(root->left);
        s.push(root->right);
        while(!s.empty()) {
            TreeNode *p = s.top(); s.pop();
            TreeNode *q = s.top(); s.pop();
            if (!p && !q) continue;
            if (!p || !q) return false;
            if (p->val != q->val) return false;
            s.push(p->left);
            s.push(q->right);
            s.push(p->right);
            s.push(q->left);
        }
        return true;
    }
};

File: Algo/Leetcode/101SymmetricTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True

        if root.left is None and root.right is None:
            return True

        return self.isSymmetricPair(root.left, root.right)

    def isSymmetricPair(self, l, r):
        if l is None and r is None:
            return True

        if l is None or r is None:
            return False

        if l.val != r.val:
            return False

        return self.isSymmetricPair(l.left, r.right) and self.isSymmetricPair(l.right, r.left)
